### DISCUSSION TITLE
[USM] Will `sycl::aligned_alloc_host` allocate pinned host momery?

### DISCUSSION THREAD

**Discussion by GHGmc2** [intel/llvm#4564] [Q&A]
- code example
```c++
#include "CL/sycl.hpp"

int main() {
  sycl::queue q(sycl::gpu_selector{});
  void *h = sycl::aligned_alloc_host(64, 256, q);
  sycl::free(h, q);
  return 0;
}
```

I notice that there is an extension for sycl buffer ["Use Pinned Host Memory Property"](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/UsePinnedMemoryProperty/UsePinnedMemoryPropery.adoc), but how about USM host allocator, will `sycl::aligned_alloc_host` allocate pinned host momery?

**Comment by GHGmc2** [comment]
A clue from Level Zero spec on ["zeMemAllocHost"](https://spec.oneapi.io/level-zero/latest/core/api.html#zememallochost):
 - "Host allocations are frequently used as staging areas to transfer data to or from devices."

**Reply by zjin-lcf** [reply]
Thank you for providing the example. Running the example with Nvprof shows pinned host memory.  I think this example also shows that the CUDA backend adds a lot of other runtime API calls. Do you think they are runtime overhead ?

No kernels were profiled.
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
      API calls:   70.95%  179.94ms         1  179.94ms  179.94ms  179.94ms  cuCtxCreate
                   28.76%  72.943ms         1  72.943ms  72.943ms  72.943ms  cuCtxDestroy
                    0.16%  405.47us         1  405.47us  405.47us  405.47us  cuMemAllocHost
                    0.10%  247.62us         1  247.62us  247.62us  247.62us  cuMemFreeHost
                    0.01%  15.124us         1  15.124us  15.124us  15.124us  cuStreamCreate
                    0.01%  14.804us         1  14.804us  14.804us  14.804us  cuStreamDestroy
                    0.00%  11.401us         6  1.9000us  1.0600us  4.6500us  cuDeviceGetPCIBusId
                    0.00%  10.350us        42     246ns     201ns     504ns  cuDeviceGetAttribute
                    0.00%  9.6150us         1  9.6150us  9.6150us  9.6150us  cuEventRecord
                    0.00%  8.1220us         1  8.1220us  8.1220us  8.1220us  cuStreamSynchronize
                    0.00%  7.9980us         1  7.9980us  7.9980us  7.9980us  cuCtxSynchronize
                    0.00%  3.7810us         1  3.7810us  3.7810us  3.7810us  cuPointerGetAttributes
                    0.00%  3.2670us        12     272ns     210ns     481ns  cuDeviceGet
                    0.00%  2.7630us         7     394ns     243ns     644ns  cuCtxGetCurrent
                    0.00%  2.2660us         1  2.2660us  2.2660us  2.2660us  cuEventCreate
                    0.00%  1.9570us         1  1.9570us  1.9570us  1.9570us  cuEventDestroy
                    0.00%  1.2620us         3     420ns     303ns     611ns  cuDeviceGetCount
                    0.00%     888ns         1     888ns     888ns     888ns  cuCtxPopCurrent
~

**Comment by GHGmc2** [comment]
@zjin-lcf Just to confirm if SYCL supports allocating pinned host memory, since sometimes it brings much benefit when doing memcpy like CUDA docs say: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#page-locked-host-memory

**Comment by zjin-lcf** [comment]
Ok. You may write a SYCL program to confirm my answer using the example in https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/. That program will be great.
